platform/engineering/design-docs/separate-content-build.md

```diff
@@ -152,56 +152,65 @@ The current front end build will be split up into two distinct builds:
 1. Content build
     - Input: CMS content
     - Output: HTML files, assets from the CMS (PDFs, images and the like)
+    - This maps pretty cleanly to the current partial deploy
 1. Application build
     - Input: `vets-website` code
     - Output: JavaScript and CSS bundles
+    - This essentially maps to the current full deploy minus the content
 
-The output of both these builds will be deployed to separate buckets. The
-reverse proxies will route traffic intelligently to the appropriate bucket given
-the request.
-
-![Proposed traffic flow to S3 buckets](images/proposed-traffic-flow-to-s3-buckets.png)
-
-**Important note:** JavaScript application landing pages will live in the CMS.
-To get an application into production, the page in Drupal will need to be
-published.
+The output of both these builds will be deployed to a single S3 bucket with
+safeguards in place to ensure they don't override each other.
 
 **Another important note:** There is no way to coordinate these two deploys to
 make an application live for the first time. The process will be to manually:
 1. Verify the application assets are live in production
-1. Publish the landing page
+1. Make the landing page live in production in the content build
+    - Depending on the answer to [this
+      question](#react-application-landing-pages), this may mean publishing the
+      page in Drupal or switching the flag in the JSON file
 1. Deploy a new content build
 
 ## Specifics
```

Comment:
I feel like there's an open question here around... how do you know which content + application builds are compatible? If content + application can be pushed separately from another, what kind of drift is possible, and how do we mitigate that?

Related q: if in an emergency, you have to roll back to a "known good release", how do you know which one that is, if you now have two separate released units that work together in tandem to serve the site?

Reply:
Yeah, good question. So the only relationships between the two builds that I can think of right now are:
- The application landing pages being present altogether
- The content on those landing pages
  - `<script>` tags
  - The DOM nodes that the React applications look for to mount on
  - This also is applicable to pages with "widgets;" they'll also need the `<script>` with the right `src` and DOM node to mount on

## Types of drift

### Missing application pages
If we had to revert a content release for whatever reason, we _may_ lose an application page.

### Contents of an application landing page change
The contents of the pages shouldn't change between releases. (I have a `todo` item to add a section on changing the cache-busting mechanism from adding hashes to bundle names to using a cache control header.)

### Rolling back a newly-released application's code, but not landing page
If we release a new application and had to roll the application deployment back, we could end up with a landing page that points to a JS bundle that doesn't exist.

### Landing page deployed before code
The content release which includes a newly prod-facing application landing page is deployed before the code is deployed to production.

I addressed this in the document; I think it'll have to be up to the teams releasing a new application to ensure that the code is live before they make the page live. I'd love to hear alternatives to this, though!

## Possible solutions
- Checking applications manually
  - Maybe not as bad an option as it sounds
  - We should know when an application page has been added and the content release has been reverted (instead of a new release deployed to fix whatever problem occurred)
  - We should also know when the application page or template changed
- Setting up an out-of-band deploy-validation test
  - What I'm envisioning now is some kind of automated browser test that loads up each page we expect to have `vets-website` code served at
  - An idea I had to serve as something of a bridge between the two systems is a kind of "application registry" repo
    - It would basically just be a single JSON file with information about each application such as the URL it should live at and whether it should be available in production (we can add whatever else we need later)
    - We could use this to tell that deployment tester where to look for each application
  - Caveat: This only tests _after_ the deploy, so it won't _prevent_ a "bad" deployment
- What else?

## Post script
This kind of drift is technically possible now (though we'd have to like change the application landing page template to not include that DOM node or something to make it happen). We currently have no way of automatically checking, so anything we do here is an enhancement.

Reply:
That was a lot of words! How much of this should I add to the document?

Reply:
most/all of it? answers to questions like this on a design doc should generally be "I've updated the doc with the information you requested" :)

Reply:
>how do you know which content + application builds are compatible?

We don't know this. We won't know this. The CMS and application builds should be completely separated. Apps teams should explicitly manage the elements in the CMS that their react applications hook to. If the elements aren't there or don't work, then it has to be fixed in the CMS. 

If we want monitoring in place to make sure that someone didn't break an app by changing something in the CMS, then that should be handled as a monitoring concern. to me this is a CMS issue- if an element in the content is driving an application, then there should be some sort of locking or alerting when that changes.

Reply:
>if in an emergency, you have to roll back to a "known good release", how do you know which one that is, if you now have two separate released units that work together in tandem to serve the site?

We can only roll back to a known good application release. My understanding is that the CMS is going to be deploying content up to every 5 minutes and could have hundreds of editors working in it at a time. We need to isolate the application deploy from this as much as possible.

Reply:
I've added this as a risk with notes about how what we're doing now.

Reply:
> We don't know this. We won't know this. The CMS and application builds should be completely separated. Apps teams should explicitly manage the elements in the CMS that their react applications hook to. If the elements aren't there or don't work, then it has to be fixed in the CMS.

One thing I've seen done for this is a policy around waiting for _two_ prod releases before referencing something new.

i.e. if an application team checks in a new React element Monday am, it goes to prod Monday pm, and then they reference it in CMS Monday evening... that puts them in danger of an application-build LKGR rollback breaking their application.

but if we have a policy that says: check in new React element Monday am, wait for TWO releases to go out before referencing it in CMS, then they should still be safe from a LKGR rollback change.

Reply:
>check in new React element Monday am, wait for TWO releases to go out before referencing it in CMS, then they should still be safe from a LKGR rollback change.

@gunsch this sounds good. Flipper doesn't support static content changes so maybe whenever we figure that out, it can smooth out this process.

Reply:
Does it make sense to outline that policy in this document?

Reply:
yeah, I think so --- at least as a proposal. no need to articulate the full explicit policy here or pretend that this design doc is a policy doc, but at least saying "we acknowledge this risk, and this is how we intend to address it" would be worth including.